File: compliance/python/workflows/parallel_fanout.py
"""
Parallel Fanout Compliance Workflow (Python reference implementation).

Parallel branch workflow: validates input, then fans out to three parallel
processing branches (enrich, score, tag), then aggregates results.

Usage:
    python parallel_fanout.py > ../tests/fixtures/parallel_fanout.json
"""

import json


def generate_expected_operations():
    """Generate the operation sequence that this workflow produces.

    The Python durable Lambda SDK records each operation as:
    - type: "step" for ctx.step() calls

    This workflow performs:
    1. validate_input — validates the incoming data
    2. enrich_data — parallel branch 1: data enrichment
    3. score_data — parallel branch 2: data scoring
    4. tag_data — parallel branch 3: data tagging
    5. aggregate_results — combines parallel branch outputs

    Note: In the Python SDK, parallel branches are implemented as
    individual steps. The Rust SDK's parallel() operation records each
    branch as a separate step operation. Both produce the same sequence.
    """
    return {
        "workflow": "parallel_fanout",
        "operations": [
            {"type": "step", "name": "validate_input"},
            {"type": "step", "name": "enrich_data"},
            {"type": "step", "name": "score_data"},
            {"type": "step", "name": "tag_data"},
            {"type": "step", "name": "aggregate_results"},
        ],
    }


if __name__ == "__main__":
    print(json.dumps(generate_expected_operations(), indent=2))

File: compliance/rust/src/parallel_fanout.rs
//! Parallel Fanout compliance workflow.
//!
//! Validates input, fans out to three parallel processing branches
//! (enrich, score, tag), then aggregates results.
//! Mirrors the Python reference implementation in `compliance/python/workflows/parallel_fanout.py`.

use durable_lambda_core::context::DurableContext;
use durable_lambda_core::error::DurableError;

/// Execute the parallel fanout workflow against the given context.
///
/// Operations recorded:
/// 1. `step:validate_input`
/// 2. `step:enrich_data`
/// 3. `step:score_data`
/// 4. `step:tag_data`
/// 5. `step:aggregate_results`
///
/// Note: The parallel branches are implemented as individual steps to match
/// the Python SDK's operation sequence. Both SDKs record each branch as a
/// separate step operation.
pub async fn run(ctx: &mut DurableContext) -> Result<serde_json::Value, DurableError> {
    // Step 1: Validate input
    let input: Result<serde_json::Value, String> = ctx
        .step("validate_input", || async {
            Ok(serde_json::json!({"data_id": "DATA-001", "valid": true}))
        })
        .await?;

    // Steps 2-4: Parallel branches (executed as individual steps for compliance)
    let enriched: Result<serde_json::Value, String> = ctx
        .step("enrich_data", || async {
            Ok(serde_json::json!({"enriched": true, "source": "external_db"}))
        })
        .await?;

    let scored: Result<serde_json::Value, String> = ctx
        .step("score_data", || async {
            Ok(serde_json::json!({"score": 0.95, "model": "v2"}))
        })
        .await?;

    let tagged: Result<serde_json::Value, String> = ctx
        .step("tag_data", || async {
            Ok(serde_json::json!({"tags": ["important", "reviewed"]}))
        })
        .await?;

    // Step 5: Aggregate results
    let aggregated: Result<serde_json::Value, String> = ctx
        .step("aggregate_results", || async {
            Ok(serde_json::json!({"status": "complete", "branches": 3}))
        })
        .await?;

    Ok(serde_json::json!({
        "input": input.unwrap(),
        "enriched": enriched.unwrap(),
        "scored": scored.unwrap(),
        "tagged": tagged.unwrap(),
        "aggregated": aggregated.unwrap(),
    }))
}
